library/Freckle/App/Kafka/Consumer.hs

```diff
@@ -0,0 +1,166 @@
+  -- ^ The offset reset parameter used when there is no initial offset in Kafka.
+  --
+  -- This is the `auto.offset.reset` Kafka consumer configuration property.
+  , kafkaConsumerConfigExtraSubscriptionProps :: Map Text Text
+  -- ^ Extra properties used to configure the Kafka consumer.
+  }
+  deriving stock (Show)
+
+envKafkaTopics
+  :: Env.Parser Env.Error (NonEmpty TopicName)
+envKafkaTopics =
+  Env.var
+    (eitherReader readKafkaTopics)
+    "KAFKA_TOPICS"
+    mempty
+
+readKafkaTopics :: String -> Either String (NonEmpty TopicName)
+readKafkaTopics t = case NE.nonEmpty $ T.splitOn "," $ T.pack t of
+  Just xs@(x NE.:| _)
+    | x /= "" -> Right $ TopicName <$> xs
+  _ -> Left "Kafka topics cannot be empty"
+
+envKafkaOffsetReset
+  :: Env.Parser Env.Error OffsetReset
+envKafkaOffsetReset =
+  Env.var
+    (eitherReader readKafkaOffsetReset)
+    "KAFKA_OFFSET_RESET"
+    $ Env.def Earliest
+
+readKafkaOffsetReset :: String -> Either String OffsetReset
+readKafkaOffsetReset t = case T.pack t of
+  "earliest" -> Right Earliest
+  "latest" -> Right Latest
+  _ -> Left "Kafka offset reset must be one of earliest or latest"
+
+envKafkaConsumerConfig
+  :: Env.Parser Env.Error KafkaConsumerConfig
+envKafkaConsumerConfig = do
+  brokerAddresses <- envKafkaBrokerAddresses
+  consumerGroupId <- Env.var Env.nonempty "KAFKA_CONSUMER_GROUP_ID" mempty
+  kafkaTopics <- envKafkaTopics
+  kafkaOffsetReset <- envKafkaOffsetReset
+  kafkaExtraProps <-
+    Env.var
+      (fmap Map.fromList . keyValues)
+      "KAFKA_EXTRA_SUBSCRIPTION_PROPS"
+      (Env.def mempty)
+  pure $
+    KafkaConsumerConfig
+      brokerAddresses
+      consumerGroupId
+      kafkaTopics
+      kafkaOffsetReset
+      kafkaExtraProps
+
+class HasKafkaConsumer env where
+  kafkaConsumerL :: Lens' env KafkaConsumer
+
+consumerProps :: KafkaConsumerConfig -> ConsumerProperties
+consumerProps KafkaConsumerConfig {..} =
+  brokersList brokers
+    <> groupId kafkaConsumerConfigGroupId
+    <> noAutoCommit
+    <> logLevel KafkaLogInfo
+ where
+  brokers = NE.toList kafkaConsumerConfigBrokerAddresses
+
+subscription :: KafkaConsumerConfig -> Subscription
+subscription KafkaConsumerConfig {..} =
+  topics (NE.toList kafkaConsumerConfigTopics)
+    <> offsetReset kafkaConsumerConfigOffsetReset
+    <> extraSubscriptionProps kafkaConsumerConfigExtraSubscriptionProps
+
+withKafkaConsumer
+  :: MonadUnliftIO m
+  => KafkaConsumerConfig
+  -> (KafkaConsumer -> m a)
+  -> m a
+withKafkaConsumer config = bracket newConsumer closeConsumer
+ where
+  (props, sub) = (consumerProps &&& subscription) config
+  newConsumer = either throwIO pure =<< Kafka.newConsumer props sub
+  closeConsumer = maybe (pure ()) throwIO <=< Kafka.closeConsumer
+
+timeoutMs :: Timeout -> Int
+timeoutMs = \case
+  TimeoutSeconds s -> s * 1000
+  TimeoutMilliseconds ms -> ms
+
+runConsumer
+  :: ( MonadUnliftIO m
+     , MonadReader env m
+     , MonadLogger m
+     , HasKafkaConsumer env
+     , FromJSON a
```

Comment:
I wonder how often we would subscribe to more than one topic which happen to have the same underlying `FromJSON a`? 

It seems like we may want to take something like `TopicName -> Value -> a -> m ()` instead of just `a -> m ()` but that feels like trying to support a need we just might not have. 

I wonder if we should make `kafkaConsumerConfigTopics :: TopicName` instead?

Reply:
I'm not sure if there is a change being suggested, but I want to clarify that `runConsumer` uses the `Subscription` that was setup in `withKafkaConsumer -> KafkaConsumerConfig -> ...` and so the first example above is actually:

```haskell
config :: KafkaConsumerConfig
config = KafkaConsumerConfig brokers groupId topics Earliest mempty
 where
  topics = NE.fromList [topicA, topicB]

loadApp :: (App -> IO a) -> IO a
loadApp f = do
  -- ...
  withKafkaConsumer config $ \consumer -> f $ App consumer

runConsumer $ \messageFromAorB -> do
```

I'm not sure how the second example would work as I think it entails having two different consumers. The first example is what I believe follows from the library because of how the `Subscription` cannot be changed/modified in a consumer that is setup, but the second example is how I originally thought the consumer API would work.

Reply:
My suggestion was to make `kafkaConsumerConfigTopics :: TopicName` instead as it would be pretty funky to support listening to more than one topic IMO.

Reply:
Just to clarify: my comment was just agreeing with AJ and showing examples in a simplified syntax. The intent was to show that passing `[topicA, topicB]` to a single consume (how the PR could be configured now) was not preferred or useful, and passing a single topic to multiple consumes (how AJ suggests the config be changed) is how we would most likely want to achieve a multi-topic app anyway.

Reply:
I am not clear on the utility of a single message handler processing across multiple topics, but I think this should align with the upstream library until we have a concrete reason to diverge and do our own thing.

Reply:
Fair enough!